### ISSUE TITLE
Unnecessary warning about panic at runtime

### ISSUE THREAD

**Issue by DominusCarnufex** [rust-lang/rust#43119]
Given the following minimal code…

```rust
macro_rules! remainder  {
    ($dividend:expr, $divisor:expr) => {
        if $divisor == 0 {
            0
        } else {
            $dividend % $divisor
        }
    };
}

fn main()   {
    #[allow(const_err)]
    let _ = remainder!(12, 0);
}
```

… I get the following warning…

```
warning: this expression will panic at run-time
  --> bad_macro.rs:6:13
   |
6  | /             $dividend % $divisor
7  | |         }
8  | |     };
9  | | }
...  |
12 | |     #[allow(const_err)]
13 | |     let _ = remainder!(12, 0);
   | |             ---------------^-
   | |_____________|______________|
   |               |              attempt to calculate the remainder with a divisor of zero
   |               in this macro invocation
```

… although I obviously thought about it and protected myself against it.

Is there any possibility to deactivate this warning? If not, there should be.

I use the latest nightly (9b85e1cfa 2017-07-07).

**Comment by kennytm** [comment]
Why not use [`$dividend.checked_rem($divisor).unwrap_or(0)`](https://doc.rust-lang.org/std/primitive.i32.html#method.checked_rem)?

**Comment by DominusCarnufex** [comment]
Your solution is neat, thank you, but it doesn’t actually tackle the issue, which is wider than the specific case I described: I get a warning, and although I know perfectly well that it is pointless, I cannot deactivate it, like I can for most other warnings.

**Comment by oli-obk** [comment]
Did const eval gain the ability to compute conditionals like `if`? Or is the fraction promoted by itself (which would be wrong out of the reasons shown here) 

You should be able to silence the warning by putting the allow inside the macro.

The reported error's span is completely broken, too.

**Comment by oli-obk** [comment]
It's not possible in general to figure out whether an expression is reachable. E.g if you put a panic before an expression that is statically known to panic, even though it's unreachable,  the error is correct. But you might write a function which returns if e.g. an unsolved math problem is solved. We can't ever prove that the function returns, but it might be obvious to a human observer. If you call that function before your statically known to panic code, we'll need to report it, because it might get executed possibly.

That said, I assume we'll need const eval to support conditionals like `if`, which requires an RFC.

**Comment by DominusCarnufex** [comment]
> You should be able to silence the warning by putting the allow inside the macro.

This doesn’t work (the warning is still here).

```rust
#![feature(stmt_expr_attributes)]

macro_rules! remainder  {
    ($dividend:expr, $divisor:expr) => {
        if $divisor == 0 {
            0
        } else {
            #[allow(const_err)]
            $dividend % $divisor
        }
    };
}

fn main()   {
    #[allow(const_err)]
    let _ = remainder!(12, 0);
}
```

This is not possible, because you cannot put attributes on if expressions.

```rust
macro_rules! remainder  {
    ($dividend:expr, $divisor:expr) => {
        #[allow(const_err)]
        if $divisor == 0 {
            0
        } else {
            $dividend % $divisor
        }
    };
}

fn main()   {
    #[allow(const_err)]
    let _ = remainder!(12, 0);
}
```

And this doesn’t work either, the warning being still present.

```rust
macro_rules! remainder  {
    ($dividend:expr, $divisor:expr) => {
        #[allow(const_err)]
        match $divisor == 0 {
            true => 0,
            false => $dividend % $divisor
        }
    };
}

fn main()   {
    #[allow(const_err)]
    let _ = remainder!(12, 0);
}
```
